Firestore/core/test/firebase/firestore/local/lru_garbage_collector_test.h

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_TEST_FIREBASE_FIRESTORE_LOCAL_LRU_GARBAGE_COLLECTOR_TEST_H_
+#define FIRESTORE_CORE_TEST_FIREBASE_FIRESTORE_LOCAL_LRU_GARBAGE_COLLECTOR_TEST_H_
+
+#include <memory>
+#include <unordered_map>
+
+#include "Firestore/core/src/firebase/firestore/auth/user.h"
+#include "Firestore/core/src/firebase/firestore/local/query_data.h"
+#include "Firestore/core/src/firebase/firestore/local/reference_set.h"
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+#include "Firestore/core/src/firebase/firestore/model/types.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+class DocumentKey;
+class SetMutation;
+
+}  // namespace model
+
+namespace local {
+
+class LruDelegate;
+class LruGarbageCollector;
+class MutationQueue;
+class Persistence;
+class QueryCache;
+class RemoteDocumentCache;
+struct LruParams;
+
+/**
+ * A set of helper methods needed by LruGarbageCollectorTest that customize it
+ * to the specific implemetnation it is testing.
+ */
+class LruGarbageCollectorTestHelper {
+ public:
+  virtual ~LruGarbageCollectorTestHelper() = default;
+
+  /** Creates a new instance of Persistence. */
+  virtual std::unique_ptr<Persistence> MakePersistence(
+      LruParams lru_params) = 0;
+
+  /** Checks whether or not a sentinel row exists for the given key. */
+  virtual bool SentinelExists(const model::DocumentKey& key) = 0;
+};
+
+using FactoryFunc = std::unique_ptr<LruGarbageCollectorTestHelper> (*)();
+
+class LruGarbageCollectorTest : public ::testing::TestWithParam<FactoryFunc> {
+ protected:
+  LruGarbageCollectorTest();
+  ~LruGarbageCollectorTest();
+
+  /**
+   * Prepares all test members based on the given LruParams, or the defaults if
+   * none are supplied.
+   */
+  void NewTestResources();
```

Comment:
Optional: use a default parameter value and remove this overload?

Reply:
`LruParams` is a forward declared type in the context, so its default value is inaccessible.

I've been erring toward avoiding `#include`s whenever possible, but I could change policy and only avoid `#Include`s only when it has little to no effect. WDYT?